source/frontend/src/components/measurements/AreaContainer.tsx

```diff
@@ -0,0 +1,37 @@
+import { AreaForm } from './AreaForm';
+import AreaView from './AreaView';
+
+interface ViewOnlyProps {
+  landArea?: number;
+  unitCode?: string;
+
+  isEdditable?: never;
+  onChange?: never;
+}
+
+interface EditableProps {
+  landArea?: number;
+  unitCode?: string;
+
+  isEdditable: boolean;
+  onChange: (landArea: number, areaUnitTypeCode: string) => void;
+}
+
+type IAreaContainerProps = EditableProps | ViewOnlyProps;
+
+function isViewOnly(areaComponent: ViewOnlyProps | EditableProps): areaComponent is ViewOnlyProps {
+  var editable = areaComponent as EditableProps;
```

Comment:
Something funky is going on, why allow both ViewOnlyProps and EditableProps but then force the type to be an EditableProps via a cast?

Reply:
The way I understand it, I believe this is an (advanced) feature of typescript called type narrowing where the function makes an assertion of the end type based on the shape of the object passed in as a parameter. A more contrived example would be:

```tsx
function isFish(pet: Fish | Bird): pet is Fish {
  return (pet as Fish).swim !== undefined;
}
``` 

https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates